### PR TITLE
[Fusilli] Add shape and stride inference to conv_fprop node

### PR DESCRIPTION
**Pull request by sjain-stanford**
As titled. Includes
- shape/stride inference for conv_fprop
- utilities to generate channels-first, channels-last and from-stride stride orders + unit tests
- preamble on memory layouts in Fusilli (setting the stage for non-contiguous support)

### CODE REVIEW

**Comment by sjain-stanford** [comment]
> I dont see any non-contiguous layout tests.

This PR doesn't add non-contiguous layout support. That's to come in a follow-on. This merely adds shape inference for the channels-first (contiguous) layouts that we currently support. Prior to this, users would need to annotate the output shapes/strides. This makes it so that's not required.

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 1]
```diff
@@ -9,6 +9,85 @@
...
+// from its logical layout. So a non-contiguous tensor of logical shape (n, c,
+// h, w) may be stored in-memory as channels-last (NHWC) using strides (h*w*c,
+// 1, w*c, c).
+//
+// Basically, strides can be used to represent different physical (in-memory)
+// layouts while preserving the same logical layout for tensors.
+//
+// PyTorch enforces tensors to always conform to the logical channels-first
+// (NCHW) layout, and uses strides as a way to specify the exact physical
+// (in-memory) layout - contiguous (NCHW) or channels-last (NHWC). Not all
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h:34]: Correct.

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 2]
```diff
@@ -9,6 +9,85 @@
...
+//    x [NCHW + stride] -> conv [NHWC] -> y [NCHW + stride]
+//
+//  Torch:
+//    x [NHWC] -> T [NCHW] -> conv [NCHW] -> T' [NHWC] -> y [NHWC]
+//
+//  Linalg:
+//    x [NHWC] -> conv [NHWC] -> T [NCHW] -> T' [NHWC] -> y [NHWC]
+//
+//    x [NHWC] -> conv [NHWC] -> y [NHWC]
+//
```
Comment (AaronStGeorge) [inline sharkfuser/include/fusilli/attributes/tensor_attributes.h:90]: Great description! I might label this these steps with their place in the flow as well as their representation:

```suggestion
//  User input (Fusilli graph):
//    x [NCHW + stride] -> conv [NHWC] -> y [NCHW + stride]
//
//  Generated ASM (Torch dialect):
//    x [NHWC] -> T [NCHW] -> conv [NCHW] -> T' [NHWC] -> y [NHWC]
//
//  Transpose elimination Passes (Linalg dialect):
//    x [NHWC] -> conv [NHWC] -> T [NCHW] -> T' [NHWC] -> y [NHWC]
//
//    x [NHWC] -> conv [NHWC] -> y [NHWC]
//
```

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 3]
```diff
@@ -207,6 +287,88 @@ struct TensorAttrSortByName {
...
+  // So we instead gate loop entry with `i > 0`.
+  for (size_t i = numDims; i > 0; --i)
+    strideOrder[i - 1] = order++;
+  return strideOrder;
+}
+
+// Generates stride order for a channels-last tensor. For a 4D tensor, this
+// would return {N: 3, C: 0, H: 2, W: 1} to represent an NHWC in-memory layout.
+// Here N is the slowest changing and C is the fastest changing dimension.
+inline std::vector<int64_t> getChannelsLastStrideOrder(size_t numDims) {
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h:312]: I'm basing this off PyTorch/cuDNN convention for spatial layouts which places the "channel" (aka "feature") dimension at dim `1` for channels-first and at dim `N-1` for channels-last, where `N` is number of dims. This applies generally for any N>=3D spatial layouts (3D, 4D, 5D...) and is not hardcoded for N,C,H,W.

For 3D, it's (N, C, L) vs (N, L, C). 
For 4D, it's (N, C, H, W) vs (N, H, W, C)
For 5D, it's (N, C, D, H, W) vs (N, D, H, W, C).

https://docs.pytorch.org/docs/stable/generated/torch.nn.Conv1d.html
https://docs.nvidia.com/deeplearning/cudnn/frontend/latest/developer/core-concepts.html#tensor-descriptor

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 4]
```diff
@@ -207,6 +287,88 @@ struct TensorAttrSortByName {
...
+                      const std::vector<int64_t> &strideOrder) {
+  size_t numDims = dim.size();
+  std::vector<int64_t> stride(numDims);
+  std::vector<std::pair<size_t, int64_t>> idxToDimInStrideOrder(numDims);
+
+  // Example dim = (10, 3, 12, 12) and strideOrder = {3, 0, 2, 1}
+  // idxToDimInStrideOrder will be:
+  // { (1, 3), (3, 12), (2, 12), (0, 10) }
+  // ordered from fastest changing dim to slowest changing dim
+  for (size_t i = 0; i < numDims; ++i)
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h:357]: The second loop depends on the fully constructed `idxToDimInStrideOrder`, which is constructed out of order (due to the double indexing -> `idxToDimInStrideOrder[strideOrder[i]]`. 
```c++
  for (size_t i = 0; i < numDims; ++i) {
    idxToDimInStrideOrder[strideOrder[i]] = std::make_pair(i, dim[i]);
  }
  for (size_t i = 0; i < numDims; ++i) {
    stride[idxToDimInStrideOrder[i].first] = cumulativeProduct;
    cumulativeProduct *= idxToDimInStrideOrder[i].second;
  }
```
Besides, I didn't try too hard to combine them as this was already getting hard to read :)
Can you share a snippet of what you mean?

File: sharkfuser/include/fusilli/node/conv_node.h [hunk 5]
```diff
@@ -64,26 +65,40 @@ class ConvFPropNode : public NodeCRTP<ConvFPropNode> {
...
-      FUSILLI_RETURN_ERROR_IF(true, ErrorCode::NotImplemented,
-                              "ConvFProp node shape inference not implemented "
-                              "yet; please specify output tensor dimensions");
+      yDim.resize(xDim.size());
+      // N
+      yDim[0] = xDim[0];
+      // K
+      yDim[1] = wDim[0];
+      // PQ...
+      for (size_t i = 2; i < xDim.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/node/conv_node.h:88]: Not sure I follow. This code doesn't rely on NCHW specifically. Those are optional dims. It should work for all spatial variants of conv - conv1d, 2d, 3d etc. So this is only as constrained as the semantics of convolution allow (e.g. this won't work for matmul but that's OK since this is specific to the ConvFProp node).

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 6]
```diff
@@ -207,6 +287,88 @@ struct TensorAttrSortByName {
...
+         "Channels-first layout requires at least 3 dimensions");
+
+  std::vector<int64_t> strideOrder(numDims);
+  int64_t order = 0;
+  // Caution: Reverse iteration with size_t (unsigned) can lead to underflow
+  // when `i == 0` is decremented due to wrap-around to SIZE_MAX (which is
+  // positive and doesn't prevent control from entering the loop as intended).
+  // This can cause heap corruption when accessing strideOrder[SIZE_MAX].
+  // So we instead gate loop entry with `i > 0`.
+  for (size_t i = numDims; i > 0; --i)
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h:304]: PyTorch calls them `torch.channels_last` and `torch.contiguous_format` (which is an alias for channels_first). I am fine changing this to `getContiguousStrideOrder` to keep it generic.

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h:358]: Ah I see what you are saying. The first loop is essentially computing the inverse, so you need to do that separately. Thanks!

File: sharkfuser/include/fusilli/node/conv_node.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply sharkfuser/include/fusilli/node/conv_node.h:88]: What I am saying is you could have
```
/// These two values could be class private values/methods
int inputWindowStartDim = 2;
int inputWindowEndDim = xDim.size()

for (size_t i = inputWindowStartDim; i < inputWindowEndDim ; ++i) {
  yDim[i] = (xDim[i]  + 2 *convFPropAttr.getPadding()[i - inputWindowStartDim] ....)
}

The value `2` is a "magic number" that Is useful to define once and use opaquely everywhere. 

```

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h (hunk 2 above)
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h]: Done

**Comment by MaheshRavishankar** [comment]
LGTM! Please land when approved by @AaronStGeorge  who has more context.

File: sharkfuser/include/fusilli/node/conv_node.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/node/conv_node.h:88]: Ah I see what you're saying. I've got a `constexpr size_t kSpatialStartIndex = 2` at the top of the method. Even if we don't end up changing the logical layouts to have multiple batch dims for a long time, this is cleaner as it avoids the repetition of the value `2` four times so thank you :)

File: sharkfuser/include/fusilli/node/conv_node.h [hunk 7]
```diff
@@ -64,26 +66,46 @@ class ConvFPropNode : public NodeCRTP<ConvFPropNode> {
...
+      // K (channel dim)
+      yDim[1] = wDim[0];
+      // PQ... (spatial dims)
+      for (size_t i = kSpatialStartIdx; i < xDim.size(); ++i) {
+        yDim[i] =
+            (xDim[i] + 2 * convFPropAttr.getPadding()[i - kSpatialStartIdx] -
+             (wDim[i] - 1) * convFPropAttr.getDilation()[i - kSpatialStartIdx] -
+             1) /
+                convFPropAttr.getStride()[i - kSpatialStartIdx] +
+            1;
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/node/conv_node.h]: Done
